pkg/controller/database/cloudsql.go

```diff
@@ -164,6 +164,15 @@ func (c *cloudsqlExternal) Create(ctx context.Context, mg resource.Managed) (man
 		return managed.ExternalCreation{}, errors.New(errNotCloudSQL)
 	}
 	cr.SetConditions(v1alpha1.Creating())
+	// NOTE(muvaf): we need to save the status, otherwise normal Update overrides
+	// it with what it sees in the api-server.
+	if err := c.kube.Status().Update(ctx, cr); err != nil {
+		return managed.ExternalCreation{}, err
+	}
+	cloudsql.AddExternalTags(cr)
+	if err := c.kube.Update(ctx, cr); err != nil {
+		return managed.ExternalCreation{}, err
+	}
```

Comment:
Could we instead call `AddExternalTags` as part of `cloudsql.GenerateDatabaseInstance`?

One of the original design principles of the managed reconciler was that `ExternalClient` implementations would not need a `client.Client`. The separation of concerns was that the reconciler interacted with the Kubernetes API and the `ExternalClient` interacted with the cloud provider API.

This separation of concerns was weakened once we started late initializing things it became common for the `Observe` method to use a client to late init the CR, but still we could be sure only the managed reconciler was updating the CR's status. I'd prefer not to start leaking that responsibility into ExternalClient implementations too, if we can avoid it.

Reply:
> Could we instead call AddExternalTags as part of cloudsql.GenerateDatabaseInstance?

A similar separation of concerns exists here too; `client` functions work merely on `ForProvider` section of the spec which describes the actual cloud provider object. I agree with the rest of the comments.

Yesterday, I thought like about an hour about where to call it and settled on this without comfort at all. Today, I found a much better place for it in 5 minutes: `Initializer`.
```
// A Initializer establishes ownership of the supplied Managed resource.
// This typically involves the operations that are run before calling any
// ExternalClient methods.
```

We use external naming initializer and I think external tagging is also a metadata operation which could be considered as an initializer. The difference is that tagging has to be implemented for each managed resource separately. Updated the PR.

Reply:
That makes perfect sense!